lib/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,272 @@
+				$template->description    = isset( $args['description'] ) ? $args['description'] : '';
+				$template->status         = 'publish';
+				$template->has_theme_file = true;
+				$template->origin         = 'plugin';
+				$template->is_custom      = true;
+				$template->post_types     = isset( $args['post_types'] ) ? $args['post_types'] : '';
+			}
+
+			$this->registered_block_templates[ $template_name ] = $template;
+
+			return $template;
+		}
+
+		/**
+		 * Retrieves all registered block templates.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @return WP_Block_Template[]|false Associative array of `$block_template_name => $block_template` pairs.
+		 */
+		public function get_all_registered() {
+			return $this->registered_block_templates;
+		}
+
+		/**
+		 * Retrieves a registered template by its and name.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string $template_name Block template name including namespace.
+		 * @return WP_Block_Template|null|false The registered block template, or null if it is not registered.
+		 */
+		public function get_registered( $template_name ) {
+			if ( ! $this->is_registered( $template_name ) ) {
+				return null;
+			}
+
+			return $this->registered_block_templates[ $template_name ];
+		}
+
+		/**
+		 * Retrieves a registered template by its slug.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string $template_slug Slug of the template.
+		 * @return WP_Block_Template|null The registered block template, or null if it is not registered.
+		 */
+		public function get_by_slug( $template_slug ) {
+			$all_templates = $this->get_all_registered();
+
+			if ( ! $all_templates ) {
+				return null;
+			}
+
+			foreach ( $all_templates as $template ) {
+				if ( $template->slug === $template_slug ) {
+					return $template;
+				}
+			}
+
+			return null;
+		}
+
+		/**
+		 * Retrieves registered block templates matching a query.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param array  $query {
+		 *     Arguments to retrieve templates. Optional, empty by default.
+		 *
+		 *     @type string[] $slug__in     List of slugs to include.
+		 *     @type string[] $slug__not_in List of slugs to skip.
+		 *     @type string   $post_type    Post type to get the templates for.
+		 * }
+		 */
+		public function get_by_query( $query = array() ) {
+			$all_templates = $this->get_all_registered();
+
+			if ( ! $all_templates ) {
+				return array();
+			}
+
+			$query            = wp_parse_args(
+				$query,
+				array(
+					'slug__in'     => array(),
+					'slug__not_in' => array(),
+					'post_type'    => '',
+				)
+			);
+			$slugs_to_include = $query['slug__in'];
+			$slugs_to_skip    = $query['slug__not_in'];
+			$post_type        = $query['post_type'];
+
+			foreach ( $all_templates as $template_name => $template ) {
+				if ( ! empty( $slugs_to_include ) && ! in_array( $template->slug, $slugs_to_include, true ) ) {
+					unset( $all_templates[ $template_name ] );
```

Comment:
Should we `unset` here? Does this affect the `registered_block_templates` reference? PHP question? Alternatively we can just build the returned array.

Reply:
I refactored it in a8f7664239312ff5a019dba19a524e4296ecf194. :slightly_smiling_face: